File: src/main/GO/leetcode/lists/offer/Offer59_2.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/9 23:11
 */

type MaxQueue struct {
	q []int
	d []int
}

func ConstructorMaxQueue() MaxQueue {
	return MaxQueue{
		q: make([]int, 0),
		d: make([]int, 0),
	}
}

func (this *MaxQueue) Max_value() int {
	if len(this.d) == 0 {
		return -1
	}
	return this.d[0]
}

func (this *MaxQueue) Push_back(value int) {
	for len(this.d) > 0 && this.d[len(this.d)-1] < value {
		this.d = this.d[:len(this.d)-1]
	}
	this.d = append(this.d, value)
	this.q = append(this.q, value)
}

func (this *MaxQueue) Pop_front() int {
	if len(this.q) == 0 {
		return -1
	}
	ans := this.q[0]
	this.q = this.q[1:]
	if ans == this.d[0] {
		this.d = this.d[1:]
	}
	return ans
}

File: src/main/java/leetcode/lists/offer/Offer59_2.java
package leetcode.lists.offer;

import java.util.Deque;
import java.util.LinkedList;
import java.util.Queue;

/**
 * @author jingxinwu
 * @date 2021-11-25 11:10 下午
 */
public class Offer59_2 {

    Queue<Integer> q;
    Deque<Integer> d;

    public Offer59_2() {
        q = new LinkedList<>();
        d = new LinkedList<>();
    }

    public int max_value() {
        if (d.isEmpty()) {
            return -1;
        }
        return d.peekFirst();
    }

    public void push_back(int value) {
        while (!d.isEmpty() && d.peekLast() < value) {
            d.pollLast();
        }
        d.offerLast(value);
        q.offer(value);
    }

    public int pop_front() {
        if (q.isEmpty()) {
            return -1;
        }
        int ans = q.poll();
        if (ans == d.peekFirst()) {
            d.pollFirst();
        }
        return ans;
    }
}
